ngrep.c

```diff
@@ -913,9 +927,35 @@ void dump_packet(struct pcap_pkthdr *h, u_char *p, uint8_t proto, unsigned char
     if (len > limitlen)
         len = limitlen;
 
-    if ((len > 0 && match_func(data, len, &match_index, &match_size) == invert_match) && !keep_matching)
+    int notmatched = len > 0 && match_func(data, len, &match_index, &match_size) == invert_match;
+    if (notmatched && !keep_matching)
         return;
 
+    if (!notmatched && reset_count)
+        follow_ip_src = NULL;
+    
+    if (keep_matching && follow_stream) {
+        if (follow_ip_src == NULL) {
+            follow_ip_src = ip_src;
```

Comment:
How much have you tested this functionality?  

You introduce the global pointer `follow_ip_src` and set it to `ip_src`, which itself is a stack-based variable in the caller `process()` - thus `ip_src` deallocates with every call.  This would make for a dangling global pointer on the next call, possibly creating a crash or a secvuln.  I can see it maybe working anyway, either due to things just lining up with the same stack address for `ip_src`, or subsequent matches still happening due to a port match. 

Do I misunderstand?